### PR TITLE
Added the Devise setup and user-authentication.

### PR DESCRIPTION
**Pull request by nshelokar-vf**
In this branch user and books tables are associated, added login-signup-logout pages and added user-authentication to create new book in the database.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -31,3 +31,6 @@
 
 # Ignore master key for decrypting credentials and more.
 /config/master.key
+
+
```
Comment (omi-vfip) [inline .gitignore]: Remove line no 34,35 and add Empty line at the end of file

File: app/controllers/private_controller.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class PrivateController < ApplicationController
+  before_action :authenticate_user!
+  def test
+    render json: {
+      message: "This is a secret message. You are seeing it because you have successfully logged in." 
+    }
+  end
+end
```
Comment (omi-vfip) [inline app/controllers/private_controller.rb:8]: Missing EOL

File: app/controllers/users/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class Users::SessionsController < Devise::SessionsController
+  respond_to :json
+  private
+  def respond_with(resource, _opts = {})
+    render json: resource
+  end
+  def respond_to_on_destroy
+    render json: { message: "Logged out." }
+  end
+end
```
Comment (omi-vfip) [inline app/controllers/users/sessions_controller.rb:10]: EOL?

File: app/controllers/users/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+class Users::SessionsController < Devise::SessionsController
+  respond_to :json
+  private
+  def respond_with(resource, _opts = {})
+    render json: resource
+  end
+  def respond_to_on_destroy
```
Comment (omi-vfip) [inline app/controllers/users/sessions_controller.rb:7]: Add an empty line before every action.

File: app/models/book.rb [hunk 5]
```diff
@@ -1,2 +1,6 @@
 class Book < ApplicationRecord
+    belongs_to :user
```
Comment (omi-vfip) [inline app/models/book.rb:2]: Use 2 space indentation for all ruby files.

File: app/models/jwt_denylist.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+class JwtDenylist < ApplicationRecord
+    include Devise::JWT::RevocationStrategies::Denylist
+    self.table_name = 'jwt_denylist'
+  end
```
Comment (omi-vfip) [inline app/models/jwt_denylist.rb:4]: indentation? EOL?

File: config/routes.rb [hunk 7]
```diff
@@ -1,11 +1,21 @@
...
 
   namespace :api do
     namespace :v1 do
       resources :books
     end
   end
-  
-  
+
+  get 'private/test'
```
Comment (omi-vfip) [inline config/routes.rb:20]: Remove this if not in use.

File: frontend/src/components/Login.jsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
+import { useRef } from "react";
+import './Signup.css';
+import { ToastContainer, toast } from 'react-toastify';
```
Comment (omi-vfip) [inline frontend/src/components/Login.jsx:3]: Third-party packages always import on top, after react import

File: frontend/src/components/Login.jsx [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import 'react-toastify/dist/ReactToastify.css';
+
+const Login = ({setCurrUser, setShow}) =>{
```
Comment (omi-vfip) [inline frontend/src/components/Login.jsx:6]: Please fix eslint issues for this file. Indentation, Spacing etc...

File: frontend/src/components/Login.jsx [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  const formRef=useRef();
+  const login=async (userInfo, setCurrUser)=>{
+    const url="http://localhost:3000/login";
```
Comment (omi-vfip) [inline frontend/src/components/Login.jsx:9]: Declare this constant outside this component with a constant variable naming convention i.e 
``` suggestion
    const BASE_URL = 'http://localhost:3000';
    const LOGIN_URL = `/${BASE_URL}/login`;
```
and use this `LOGIN_URL`.

File: frontend/src/components/Login.jsx [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    try{
+        const response=await fetch(url, {
```
Comment (omi-vfip) [inline frontend/src/components/Login.jsx]: We can use `Axios` for HTTP requests later.

File: frontend/src/components/Login.jsx [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+      <div className="login-link">
+        Not registered yet? <a href="#signup" onClick={handleClick}>Sign up</a>.
+      </div>
+      <ToastContainer />
+    </div>
+  )
+}
+export default Login
+
+
```
Comment (omi-vfip) [inline frontend/src/components/Login.jsx]: Remove these empty lines.

File: app/controllers/api/v1/books_controller.rb [hunk 13]
```diff
@@ -1,30 +1,25 @@
...
   end
-
-  # POST /books
+  
   def create
-    @book = Book.new(book_params)
-
+    @book = current_user.books.build(book_params)
     if @book.save
       render json: @book, status: :created, location: api_v1_book_url(@book)
```
Comment (skumar-vf) [inline app/controllers/api/v1/books_controller.rb:17]: @nshelokar-vf Use [jbuilder](https://github.com/rails/jbuilder) when returning json.

File: app/controllers/api/v1/books_controller.rb [hunk 14]
```diff
@@ -33,18 +28,15 @@ def update
...
   end
 
-  # DELETE /books/1
   def destroy
     @book.destroy!
   end
 
   private
-    # Use callbacks to share common setup or constraints between actions.
     def set_book
```
Comment (skumar-vf) [inline app/controllers/api/v1/books_controller.rb:36]: @nshelokar-vf Can we name it better?

File: app/controllers/api/v1/books_controller.rb [hunk 15]
```diff
@@ -1,30 +1,25 @@
 class Api::V1::BooksController < ApplicationController
+  before_action :authenticate_user!
   before_action :set_book, only: %i[ show update destroy ]
```
Comment (skumar-vf) [inline app/controllers/api/v1/books_controller.rb:3]: @nshelokar-vf Don't use trailing or leading whitespace here inside array literals

File: app/controllers/api/v1/books_controller.rb [hunk 16]
```diff
@@ -1,30 +1,25 @@
...
-
-  # GET /books
+  
   def index
-    @books = Book.all
-
+    @books = current_user.books
```
Comment (skumar-vf) [inline app/controllers/api/v1/books_controller.rb:6]: @nshelokar-vf Why are initialising with instance variable here? Are we using this elsewhere outside the method scope?

File: app/controllers/api/v1/books_controller.rb [hunk 17]
```diff
@@ -1,30 +1,25 @@
...
   def show
     render json: @book
   end
-
-  # POST /books
+  
   def create
-    @book = Book.new(book_params)
-
+    @book = current_user.books.build(book_params)
```
Comment (skumar-vf) [inline app/controllers/api/v1/books_controller.rb:15]: @nshelokar-vf unnecessary declaration of instance variable.

File: .gitignore (hunk 1 above)
Comment (nshelokar-vf) [reply .gitignore]: Done sir.

File: frontend/src/App.css [hunk 18]
```diff
@@ -1,8 +1,24 @@
+html, body, #root, .App {
+  height: 100%;
+  margin: 0;
+  width: 100%;
+  padding: 0;
+}
+
 #root {
-  max-width: 1280px;
+  height: 100%;
```
Comment (omi-vfip) [inline frontend/src/App.css:9]: is there any specific reason for adding this `height: 100%;` to the `#root` element id?

File: frontend/src/App.jsx [hunk 19]
```diff
@@ -1,23 +1,12 @@
...
-    <Router>
-      <div className='app'>
-        <h2>Book Library</h2>
-        <NavBar/>
-        <AppRoutes/>
-      </div>
-    </Router>
-  
-    </>
+    <div className='App'>
```
Comment (omi-vfip) [inline frontend/src/App.jsx:6]: Use double quotation marks to assign string values to element. 
```suggestion
    <div className="App">
```

File: frontend/src/features/books/NewBook.css [hunk 20]
```diff
@@ -0,0 +1,58 @@
+.form-container {
```
Comment (omi-vfip) [inline frontend/src/features/books/NewBook.css:1]: Using snake_case for CSS file names is a common practice

File: frontend/src/features/books/CreateBook.jsx [hunk 21]
```diff
@@ -1,43 +1,53 @@
...
-import { useNavigate } from 'react-router-dom';
+import { useState, useEffect } from "react";
+import { useNavigate, Link } from 'react-router-dom';
+import { ToastContainer, toast } from 'react-toastify';
+import 'react-toastify/dist/ReactToastify.css';
 import { API_URL } from "../../constants";
+import './NewBook.css';
 
-function NewPostForm() {
+function CreateBook() {
```
Comment (omi-vfip) [inline frontend/src/features/books/CreateBook.jsx:8]: Use the arrow function from now onwards

File: frontend/src/features/books/BookList.jsx [hunk 22]
```diff
@@ -9,56 +10,67 @@ function BookList() {
 
   useEffect(() => {
     async function loadBooks() {
+      const token = localStorage.getItem("token");
+      if (!token) {
+        setError("You need to sign in or sign up before continuing.-nupur");
```
Comment (omi-vfip) [inline frontend/src/features/books/BookList.jsx:15]:
```suggestion
        setError("You need to sign in or sign up before continuing.");
```

**Comment by nshelokar-vf** [comment]
Merging this PR to `main` branch. Review comments in this PR will be addressed in Edit Feature PR i.e https://github.com/nshelokar-vf/library_management/pull/2

File: frontend/src/features/books/CreateBook.jsx (hunk 21 above)
Comment (nshelokar-vf) [reply frontend/src/features/books/CreateBook.jsx:8]: Done sir.

File: frontend/src/App.jsx (hunk 19 above)
Comment (nshelokar-vf) [reply frontend/src/App.jsx:6]: Okay sir.
Done.

File: frontend/src/App.css (hunk 18 above)
Comment (nshelokar-vf) [reply frontend/src/App.css:9]: I added it while trying to add css to the page, but now i have removed it.

File: app/controllers/api/v1/books_controller.rb (hunk 17 above)
Comment (nshelokar-vf) [reply app/controllers/api/v1/books_controller.rb:15]: Removed it.

File: app/controllers/api/v1/books_controller.rb (hunk 16 above)
Comment (nshelokar-vf) [reply app/controllers/api/v1/books_controller.rb:6]: No sir, not using it elsewhere. Removed it now.

File: app/controllers/api/v1/books_controller.rb (hunk 15 above)
Comment (nshelokar-vf) [reply app/controllers/api/v1/books_controller.rb:3]: Okay sir.

File: app/controllers/api/v1/books_controller.rb (hunk 14 above)
Comment (nshelokar-vf) [reply app/controllers/api/v1/books_controller.rb:36]: Can I name it find_book? or please suggest me a proper name.

File: frontend/src/components/Login.jsx (hunk 11 above)
Comment (nshelokar-vf) [reply frontend/src/components/Login.jsx]: Yes Done sir.

File: frontend/src/components/Login.jsx (hunk 8 above)
Comment (nshelokar-vf) [reply frontend/src/components/Login.jsx:3]: Okay Sir.

File: config/routes.rb (hunk 7 above)
Comment (nshelokar-vf) [reply config/routes.rb:20]: Sure Sir.

File: app/models/jwt_denylist.rb (hunk 6 above)
Comment (nshelokar-vf) [reply app/models/jwt_denylist.rb:4]: Done Sir.

File: frontend/src/components/Login.jsx (hunk 10 above)
Comment (nshelokar-vf) [reply frontend/src/components/Login.jsx:9]: Okay Done Sir.

File: app/controllers/private_controller.rb (hunk 2 above)
Comment (nshelokar-vf) [reply app/controllers/private_controller.rb:8]: Added it.

File: frontend/src/components/Login.jsx (hunk 9 above)
Comment (nshelokar-vf) [reply frontend/src/components/Login.jsx:6]: Done sir.

**Pull request merged**